File: src/oca/java/src/org/opennebula/client/host/HostPool.java
/*******************************************************************************
 * Copyright 2002-2024, OpenNebula Project, OpenNebula Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ******************************************************************************/
package org.opennebula.client.host;

import java.util.AbstractList;
import java.util.Iterator;


import org.opennebula.client.Client;
import org.opennebula.client.OneResponse;
import org.opennebula.client.Pool;
import org.opennebula.client.PoolElement;
import org.w3c.dom.Node;

/**
 * This class represents an OpenNebula host pool.
 * It also offers static XML-RPC call wrappers.
 */
public class HostPool extends Pool implements Iterable<Host>{

    private static final String ELEMENT_NAME = "HOST";
    private static final String INFO_METHOD  = "hostpool.info";
    private static final String MONITORING   = "hostpool.monitoring";

    /**
     * Creates a new host pool
     * @param client XML-RPC Client.
     */
    public HostPool(Client client)
    {
        super(ELEMENT_NAME, client, INFO_METHOD);
    }

    @Override
    public PoolElement factory(Node node)
    {
        return new Host(node, client);
    }

    /**
     * Retrieves all the hosts in the pool.
     *
     * @param client XML-RPC Client.
     * @return If successful the message contains the string
     * with the information returned by OpenNebula.
     */
    public static OneResponse info(Client client)
    {
        return Pool.info(client, INFO_METHOD);
    }

    /**
     * Retrieves the monitoring data for all the hosts in the pool.
     *
     * @param client XML-RPC Client.
     * @return If successful the message contains the string
     * with the information returned by OpenNebula.
     */
    public static OneResponse monitoring(Client client)
    {
        return client.call(MONITORING);
    }

    /**
     * Retrieves the monitoring data for all the hosts in the pool.
     *
     * @param client XML-RPC Client.
     * @param num: Retrieve monitor records in the last num seconds.
     * 0 just the last record, -1 all records.
     * @return If successful the message contains the string
     * with the information returned by OpenNebula.
     */
    public static OneResponse monitoring(Client client, int num)
    {
        return client.call(MONITORING, num);
    }

    /**
     * Loads the xml representation of the host pool.
     *
     * @see HostPool#info(Client)
     */
    public OneResponse info()
    {
        return super.info();
    }

    /**
     * Retrieves the monitoring data for all the hosts in the pool.
     *
     * @return If successful the message contains the string
     * with the information returned by OpenNebula.
     */
    public OneResponse monitoring()
    {
        return monitoring(client);
    }

    public Iterator<Host> iterator()
    {
        AbstractList<Host> ab = new AbstractList<Host>()
        {
            public int size()
            {
                return getLength();
            }

            public Host get(int index)
            {
                return (Host) item(index);
            }
        };

        return ab.iterator();
    }

    /**
     * Returns the Host with the given Id from the pool. If it is not found,
     * then returns null. The method {@link #info()} must be called before.
     *
     * @param id of the Host to retrieve
     * @return The Image with the given Id, or null if it was not found.
     */
    public Host getById(int id)
    {
        return (Host) super.getById(id);
    }
}

File: src/oca/ruby/opennebula/host_pool.rb
# -------------------------------------------------------------------------- #
# Copyright 2002-2024, OpenNebula Project, OpenNebula Systems                #
#                                                                            #
# Licensed under the Apache License, Version 2.0 (the "License"); you may    #
# not use this file except in compliance with the License. You may obtain    #
# a copy of the License at                                                   #
#                                                                            #
# http://www.apache.org/licenses/LICENSE-2.0                                 #
#                                                                            #
# Unless required by applicable law or agreed to in writing, software        #
# distributed under the License is distributed on an "AS IS" BASIS,          #
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   #
# See the License for the specific language governing permissions and        #
# limitations under the License.                                             #
#--------------------------------------------------------------------------- #


require 'opennebula/pool'

module OpenNebula
    class HostPool < Pool
        #######################################################################
        # Constants and Class attribute accessors
        #######################################################################


        HOST_POOL_METHODS = {
            :info       => "hostpool.info",
            :monitoring => "hostpool.monitoring"
        }

        #######################################################################
        # Class constructor & Pool Methods
        #######################################################################


        # +client+ a Client object that represents a XML-RPC connection
        def initialize(client)
            super('HOST_POOL','HOST',client)
        end

        # Factory Method for the Host Pool
        def factory(element_xml)
            OpenNebula::Host.new(element_xml,@client)
        end

        #######################################################################
        # XML-RPC Methods for the Host Pool
        #######################################################################

        # Retrieves all the Hosts in the pool.
        def info()
            super(HOST_POOL_METHODS[:info])
        end

        def info_all()
            return super(HOST_POOL_METHODS[:info])
        end

        def info_mine()
            return super(HOST_POOL_METHODS[:info])
        end

        def info_group()
            return super(HOST_POOL_METHODS[:info])
        end

        alias_method :info!, :info
        alias_method :info_all!, :info_all
        alias_method :info_mine!, :info_mine
        alias_method :info_group!, :info_group

        # Retrieves the monitoring data for all the Hosts in the pool
        #
        # @param [Array<String>] xpath_expressions Elements to retrieve.
        #
        # @return [Hash<String, <Hash<String, Array<Array<int>>>>>,
        #   OpenNebula::Error] The first level hash uses the Host ID as keys,
        #   and as value a Hash with the requested xpath expressions,
        #   and an Array of 'timestamp, value'.
        #
        # @example
        #   host_pool.monitoring(
        #     ['HOST_SHARE/FREE_CPU',
        #     'HOST_SHARE/RUNNING_VMS',
        #     'TEMPLATE/CUSTOM_PROBE'] )
        #
        #   {"1"=>
        #     {"TEMPLATE/CUSTOM_PROBE"=>[],
        #      "HOST_SHARE/FREE_CPU"=>[["1337609673", "800"]],
        #      "HOST_SHARE/RUNNING_VMS"=>[["1337609673", "3"]]},
        #    "0"=>
        #     {"TEMPLATE/CUSTOM_PROBE"=>[],
        #      "HOST_SHARE/FREE_CPU"=>[["1337609673", "800"]],
        #      "HOST_SHARE/RUNNING_VMS"=>[["1337609673", "3"]]}}
        def monitoring(xpath_expressions)
            return super(HOST_POOL_METHODS[:monitoring], xpath_expressions)
        end

        # Retrieves the monitoring data for all the Hosts in the pool, in XML
        #
        # @param [Integer] num Optional Retrieve monitor records in the last num
        #   seconds. 0 just the last record, -1 or nil all records
        #
        # @return [String] VM monitoring data, in XML
        def monitoring_xml(num = nil)
            return @client.call(HOST_POOL_METHODS[:monitoring]) if num.nil?

            @client.call(HOST_POOL_METHODS[:monitoring], num.to_i)
        end

    end

end
